src/Controller/PostfinanceResponseController.php

```diff
@@ -25,27 +26,42 @@ class PostfinanceResponseController {
    * (pending, status 51 or 91).
    *
    * @param \Symfony\Component\HttpFoundation\Request $request
-   *   Request
+   *   Request.
    * @param \Drupal\payment\Entity\PaymentInterface $payment
    *   The Payment Entity type.
+   *
+   * @return \Drupal\Core\Response
+   *   The Response to the accepting request.
    */
   public function processAcceptResponse(Request $request, PaymentInterface $payment) {
     // The definition of the plugin implementation.
     $plugin_definition = $payment->getPaymentMethod()->getPluginDefinition();
+    try {
+      $request_data = $request->query->all();
+      $sha_sent = $request_data['SHASIGN'];
+      unset($request_data['SHASIGN']);
+
+      // Generate SHASign from request data.
+      $sha_sign = PostfinanceHelper::generateShaSign($request_data, $plugin_definition['sha_out_key']);
 
-    // Generate local SHASign.
-    $request_values = $request->query->all();
+      // Check if the sent signature is valid.
+      if ($sha_sign != $sha_sent) {
+        throw new \Exception('Invalid postfinance key.');
+      }
 
-    // Generate SHA-OUT signature
-    $sha_sign = PostfinanceHelper::generateShaSign($request_values, $plugin_definition['sha_out_key']);
+      if ($request_data['STATUS'] == 'error') {
+        throw new \Exception('There was an error processing the request:' . $request_data['NCERROR']);
```

Comment:
Same here, exception is not a good way to do this. Maybe a drupal_set_message() + logging, and redirecting the user to where he came from.